File: Algorithms/Sorts/Cpp/BubbleSort.cpp
#include <iostream>

using namespace std;

void swap(int *xp, int *yp);
void bubbleSort(int *arr, int n);
void printArray(int *arr, int n);

int main(){
    // sort 8, 5, 3, 1, 9, 6, 0, 7, 4, 2
    int arr[] = {8, 5, 3, 1, 9, 6, 0, 7, 4, 2};
    int n = sizeof(arr) / sizeof(arr[0]);
    bubbleSort(arr, n);
    printArray(arr, n);
    return 0;
}

void bubbleSort(int *arr, int n)
{
    for (int i = 0; i < n; i++) // go through the array n times
    {
        for (int j = 0; j < n - 1; j++) // -1 because we are comparing j and j+1
        {
            if (arr[j] > arr[j + 1])
            {
                swap(arr[j], arr[j + 1]);
            }
        }
    }
}

void swap(int *xp, int *yp)
{
    int temp = *xp;
    *xp = *yp;
    *yp = temp;
}

void printArray(int *arr, int n)
{
    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;
}

File: Algorithms/Sorts/Csharp/BubbleSort.cs
public static class BubbleSort
{
    public static void BubSort(int[] array)
    {
        bool isSwapped; // If an iteration passes without a swap, it means that the array is sorted.
        int n = array.Length;
        
        do
        {
            isSwapped = false;
            for (int i = 1; i < n; i++)
            {
                if (array[i - 1] > array[i])
                {
                    (array[i - 1], array[i]) = (array[i], array[i - 1]); // C# swapping two variables without temp variable
                    isSwapped = true;
                }
            }
            n--;
        } while (isSwapped);
    }

    public static void Main()
    {
        int[] array = new int[]{ 9, 8, 7, 5, 4, 3, 2, 1};
        
        Console.WriteLine("Unsorted array:");
        PrintArray(array);

        BubSort(array);
        
        Console.WriteLine("Sorted array:");
        PrintArray(array);

    }

    private static void PrintArray(int[] array)
    {
        for (int i = 0; i < array.Length; i++)
        {
            Console.Write($"{array[i]} ");
        }
        Console.WriteLine("");
    }
}

